src/main/java/com/medallia/word2vec/neuralnetwork/NeuralNetworkTrainer.java

```diff
@@ -151,28 +156,30 @@ static long incrementRandom(long r) {
 	}
 	
 	/** @return Trained NN model */
-	public NeuralNetworkModel train(Iterable<List<String>> sentences) throws InterruptedException {
-		ListeningExecutorService ex = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(config.numThreads));
-		
+	public NeuralNetworkModel train(final Iterable<List<String>> sentences) throws InterruptedException {
+		final ListeningExecutorService ex =
+				MoreExecutors.listeningDecorator(
+					new ThreadPoolExecutor(config.numThreads, config.numThreads,
+							0L, TimeUnit.MILLISECONDS,
+							new ArrayBlockingQueue<Runnable>(8),
+							new ThreadPoolExecutor.CallerRunsPolicy()));
+
 		int numSentences = Iterables.size(sentences);
 		numTrainedTokens += numSentences;
-		
-		// Partition the sentences evenly amongst the threads
-		Iterable<List<List<String>>> partitioned = Iterables.partition(sentences, numSentences / config.numThreads + 1);
-		
+
+		// Partition the sentences into batches
+		final Iterable<List<List<String>>> batched = Iterables.partition(sentences, 1024);
```

Comment:
It's necessary because in the old version, each of the inner lists is completely materialized in memory, all at the same time. If I have 10M sentences, and 10 processors, it makes 10 lists of 1M sentences each, and keep them in memory. Unless your entire corpus fits in memory, this will never run.

I looked at the code that updates the learning rate. It looks like the problem is that this change might affect which sentences get used when the learning rate is high, and which ones get used when it's lower? If sentences get processed in a different order, doesn't that also change the neural net, and therefore the gradients, that are encountered by each sentence?

Either way, that wasn't deterministic before, and it still isn't. I'd suspect that the number of cores affects the output as well, regardless of whether it runs the old version of this code or the new one.